portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java

```diff
@@ -2065,36 +2066,50 @@ private void _addFileEntryResources(
 			DLFileEntry dlFileEntry, ServiceContext serviceContext)
 		throws PortalException {
 
-		if (serviceContext.isAddGroupPermissions() ||
-			serviceContext.isAddGuestPermissions()) {
+		List<Long> ancestorFolderIds = new ArrayList<>();
 
-			_resourceLocalService.addResources(
-				dlFileEntry.getCompanyId(), dlFileEntry.getGroupId(),
-				dlFileEntry.getUserId(), DLFileEntry.class.getName(),
-				dlFileEntry.getFileEntryId(), false, serviceContext);
-		}
-		else {
-			if (serviceContext.isDeriveDefaultPermissions()) {
-				serviceContext.deriveDefaultPermissions(
-					dlFileEntry.getRepositoryId(),
-					DLFileEntryConstants.getClassName());
-			}
+		if (dlFileEntry.getFolderId() !=
+				DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
 
-			_resourceLocalService.addModelResources(
-				dlFileEntry.getCompanyId(), dlFileEntry.getGroupId(),
-				dlFileEntry.getUserId(), DLFileEntry.class.getName(),
-				dlFileEntry.getFileEntryId(),
-				serviceContext.getModelPermissions());
+			DLFolder folder = dlFileEntry.getFolder();
+
+			ancestorFolderIds.add(folder.getFolderId());
+			ancestorFolderIds.addAll(folder.getAncestorFolderIds());
 		}
 
 		long inheritableParentFolderId =
 			DLPermissionPropagationUtil.getInheritableParentFolderId(
-				dlFileEntry);
+				dlFileEntry.getCompanyId(), dlFileEntry.getGroupId(),
+				ancestorFolderIds);
```

Comment:
Should use `DLPermissionPropagationUtil.getInheritableParentFolderId(dlFileEntry)` instead of.